pkg/admission/permissionclaims/mutating_permission_claims.go

```diff
@@ -0,0 +1,128 @@
+	byWorkspaceIndex = "mutating-permission-byWorkspace"
+	PluginName       = "apis.kcp.dev/PermissionClaims"
+)
+
+type mutatingPermissionClaims struct {
+	*admission.Handler
+
+	apiBindingsIndexer cache.Indexer
+	bindingReady       func() bool
+}
+
+var _ admission.MutationInterface = &mutatingPermissionClaims{}
+
+func NewMutatingPermissionClaims() admission.MutationInterface {
+
+	p := &mutatingPermissionClaims{}
+	p.Handler = admission.NewHandler(admission.Create)
+	p.SetReadyFunc(p.bindingReady)
+	return p
+}
+
+func (m *mutatingPermissionClaims) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	lcluster, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		return err
+	}
+
+	bindings, err := m.apiBindingsIndexer.ByIndex(byWorkspaceIndex, lcluster.String())
+	if err != nil {
+		return err
+	}
+
+	for _, b := range bindings {
+		binding, ok := b.(*apisv1alpha1.APIBinding)
+		if !ok {
+			return fmt.Errorf("expected a certain type")
+		}
+		for _, pc := range binding.Status.ObservedAcceptedPermissionClaims {
+			if pc.Group == a.GetKind().Group && pc.Resource == a.GetResource().Resource {
+				key, label, err := permissionclaims.PermissionClaimToLabel(pc)
+				if err != nil {
+					return err
+				}
+				u, ok := a.GetObject().(metav1.Object)
+				if !ok {
+					return fmt.Errorf("expected type")
+				}
+				l := u.GetLabels()
+				if l == nil {
+					l = map[string]string{}
+				}
+				l[key] = label
+				u.SetLabels(l)
+			}
+		}
+	}
+
+	return nil
+}
+
+// SetKcpInformers implements the WantsExternalKcpInformerFactory interface.
+func (m *mutatingPermissionClaims) SetKcpInformers(f kcpinformers.SharedInformerFactory) {
+	if _, found := f.Apis().V1alpha1().APIBindings().Informer().GetIndexer().GetIndexers()[byWorkspaceIndex]; !found {
+		if err := f.Apis().V1alpha1().APIBindings().Informer().AddIndexers(cache.Indexers{
+			byWorkspaceIndex: func(obj interface{}) ([]string, error) {
+				return []string{logicalcluster.From(obj.(metav1.Object)).String()}, nil
+			},
+		}); err != nil {
+			// nothing we can do here. But this should also never happen. We check for existence before.
+			klog.Errorf("failed to add indexer for APIBindings: %v", err)
+		}
+	}
+	if _, found := f.Apis().V1alpha1().APIExports().Informer().GetIndexer().GetIndexers()[byWorkspaceIndex]; !found {
+		if err := f.Apis().V1alpha1().APIExports().Informer().AddIndexers(cache.Indexers{
+			byWorkspaceIndex: func(obj interface{}) ([]string, error) {
+				return []string{logicalcluster.From(obj.(metav1.Object)).String()}, nil
+			},
+		}); err != nil {
+			// nothing we can do here. But this should also never happen. We check for existence before.
+			klog.Errorf("failed to add indexer for APIBindings: %v", err)
+		}
+	}
+	m.apiBindingsIndexer = f.Apis().V1alpha1().APIBindings().Informer().GetIndexer()
+	m.bindingReady = f.Apis().V1alpha1().APIBindings().Informer().HasSynced
+}
+
+func Register(plugins *admission.Plugins) {
```

Comment:
Register is traditionally at the very top.

Reply:
Moved to the top!